exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/MergeJoinBatch.java

```diff
@@ -297,7 +297,14 @@ public void close() {
       batchMemoryManager.getAvgOutputRowWidth(), batchMemoryManager.getTotalOutputRecords());
 
     super.close();
-    leftIterator.close();
+    try {
+      leftIterator.close();
+    } catch (Exception e) {
```

Comment:
Do we know what kind(s) of exceptions to expect here?   Also, can we throw a better error message?   Specifically, can we tell the user more information about the cause of the crash and how to fix it?

Reply:
@cgivre 
In my test case ,it throw  "QueryCancelledException",because some minorfragment throw .OutOfMemoryException ,so it inform foreman failed.

foreman send "QueryCancel" commands to other minorfragments. it  throws QueryCancelledException after the method "incoming.next()"   called method checkContinue() method

Although the "checkContinue" phase throws a fixed "QueryCancelledException" message, I am not sure what is causing it (In my test case ,OutOfMemoryException cause exception)


 ```
 public void clearInflightBatches() {
    while (lastOutcome == IterOutcome.OK || lastOutcome == IterOutcome.OK_NEW_SCHEMA) {
      // Clear all buffers from incoming.
      for (VectorWrapper<?> wrapper : incoming) {
        wrapper.getValueVector().clear();
      }
      lastOutcome = incoming.next();
    }
  }
   
   public void checkContinue() {
      if (!shouldContinue()) {
        throw new QueryCancelledException();
      }
    }
  }
```

**stack**

```
Caused by: org.apache.drill.exec.ops.QueryCancelledException: null
        at org.apache.drill.exec.work.fragment.FragmentExecutor$ExecutorStateImpl.checkContinue(FragmentExecutor.java:533)
        at org.apache.drill.exec.record.AbstractRecordBatch.checkContinue(AbstractRecordBatch.java:278)
        at org.apache.drill.exec.record.AbstractRecordBatch.next(AbstractRecordBatch.java:105)
        at org.apache.drill.exec.record.AbstractUnaryRecordBatch.innerNext(AbstractUnaryRecordBatch.java:59)
        at org.apache.drill.exec.record.AbstractRecordBatch.next(AbstractRecordBatch.java:165)
        at org.apache.drill.exec.record.RecordIterator.clearInflightBatches(RecordIterator.java:359)
        at org.apache.drill.exec.record.RecordIterator.close(RecordIterator.java:365)
        at org.apache.drill.exec.physical.impl.join.MergeJoinBatch.close(MergeJoinBatch.java:301)
```

Reply:
add exception info  ?
```
try {
      leftIterator.close();
    } catch (QueryCancelledException qce) {
      throw UserException.executionError(qce)
          .message("Failed when depleting incoming batches, probably because query was cancelled " +
              "by executor had some error")
          .build(logger);
    } catch (Exception e) {
      throw UserException.internalError(e)
          .message("Failed when depleting incoming batches")
          .build(logger);
    } finally {
      // todo catch exception info or By default,the exception is thrown directly ?
      rightIterator.close();
    }
```

Reply:
stack 

```
Caused by: org.apache.drill.exec.ops.QueryCancelledException: null
        at org.apache.drill.exec.work.fragment.FragmentExecutor$ExecutorStateImpl.checkContinue(FragmentExecutor.java:533)
        at org.apache.drill.exec.record.AbstractRecordBatch.checkContinue(AbstractRecordBatch.java:278)
        at org.apache.drill.exec.record.AbstractRecordBatch.next(AbstractRecordBatch.java:105)
        at org.apache.drill.exec.record.AbstractUnaryRecordBatch.innerNext(AbstractUnaryRecordBatch.java:59)
        at org.apache.drill.exec.record.AbstractRecordBatch.next(AbstractRecordBatch.java:165)
        at org.apache.drill.exec.record.RecordIterator.clearInflightBatches(RecordIterator.java:359)
        at org.apache.drill.exec.record.RecordIterator.close(RecordIterator.java:365)
        at org.apache.drill.exec.physical.impl.join.MergeJoinBatch.close(MergeJoinBatch.java:301)
```